src/dcp/dcp.c

```diff
@@ -470,104 +595,160 @@ int main(int argc, char** argv)
+     * Then, perform a reduction and exit if any process errored. */
+    bool local_daos_error = false;
+    bool global_daos_error = false;
+    int flag_daos_args;
+
+    /* Make sure we have the required DAOS arguments (if any).
+     * Safe to exit here, since all processes have the same values. */
+    rc = daos_check_args(rank, argpaths, src_pool_uuid, src_cont_uuid,
+            dst_pool_uuid, dst_cont_uuid, src_svc, dst_svc, dfs_prefix,
+            &flag_daos_args);
+    if (rc != 0) {
+        if (rank == 0) {
+            MFU_LOG(MFU_LOG_ERR, "Invalid DAOS args: "
+                    MFU_ERRF, MFU_ERRP(-MFU_ERR_DAOS_INVAL_ARG));
+        }
+        mfu_finalize();
+        MPI_Finalize();
+        return 1;
+    }
+
+    /* For now, track the error.
+     * Later, ignore if no daos args supplied */
+    rc = daos_init();
     if (rc != 0) {
         MFU_LOG(MFU_LOG_ERR, "Failed to initialize daos");
+        local_daos_error = true;
     }
 
     /* Figure out if daos path is the src or dst,
      * using UNS path, then chop off UNS path
      * prefix since the path is mapped to the root
      * of the container in the DAOS DFS mount */
-    if (!daos_uuid_valid(src_pool_uuid) || !daos_uuid_valid(dst_pool_uuid)) {
-        daos_set_paths(argpaths, dfs_prefix, src_pool_uuid, src_cont_uuid,
-            dst_pool_uuid, dst_cont_uuid, mfu_src_file, mfu_dst_file);
+    if (!local_daos_error
+            && (!daos_uuid_valid(src_pool_uuid) || !daos_uuid_valid(dst_pool_uuid))) {
+        rc = daos_set_paths(rank, argpaths, dfs_prefix, src_pool_uuid, src_cont_uuid,
+                dst_pool_uuid, dst_cont_uuid, mfu_src_file, mfu_dst_file);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Invalid DAOS args: "
+                    MFU_ERRF, MFU_ERRP(-MFU_ERR_DAOS_INVAL_ARG));
+            local_daos_error = true;
+        }
     }
 
+    /* Re-check the required DAOS arguments (if any) */
+    if (!local_daos_error) {
+        rc = daos_check_args(rank, argpaths, src_pool_uuid, src_cont_uuid,
+                dst_pool_uuid, dst_cont_uuid, src_svc, dst_svc, dfs_prefix,
+                &flag_daos_args);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "Invalid DAOS args: "
+                    MFU_ERRF, MFU_ERRP(-MFU_ERR_DAOS_INVAL_ARG));
+            local_daos_error = true;
+        }
+    }
+    
     /* check if DAOS source and destination containers are in the same pool */
     bool same_pool = false;
     if (mfu_src_file->type == DAOS && mfu_dst_file->type == DAOS) {
         if (uuid_compare(src_pool_uuid, dst_pool_uuid) == 0) {
             same_pool = true;
         }
-    } 
+    }
 
     /* connect to DAOS source pool if uuid is valid */
-    if (mfu_src_file->type == DAOS) {
-        /* if DAOS source pool uuid is valid, then set source file type to DAOS */
-        daos_connect(rank, src_svc, src_pool_uuid, src_cont_uuid, &src_poh, &src_coh); 
+    if (!local_daos_error && mfu_src_file->type == DAOS) {
```

Comment:
Skimming the code, I get the impression that local_daos_error is tracking whether the calling process has seen an error, which may be independent of other processes.  Is that right?

Reply:
Correct. It is *technically possible* that some DAOS call might fail for only a single process, some of which might stem from erroneous errors. Practically speaking, most processes at the point of the broadcast into global_daos_error are *most likely* in a similar state, but to be safe in calling MPI_Finalize(), etc., it's best to make sure. 
One such out-of sync issue that could possibly arise, though unlikely, would be in our call to duns_resolve_path. 
Let's say we have 2 processes. One of them calls duns_resolve_path and succeeds. Before the other process makes the call, the UNS path is removed from the DAOS system, and then the 2nd process makes the call, and fails. Now the processes are in different states.
This is where my reasoning for "local_daos_error" comes from, among other erroneous network errors that might crop up.

Reply:
Assuming that's the case, one thing to be careful of is mixed error conditions, like if local_daos_error=true on rank=0, but false on rank=1.  It looks like daos_connect includes a collective (bcast), but then you might have rank 0 skipping that call while rank 1 does.

Reply:
> 
> 
> Correct. It is _technically possible_ that some DAOS call might fail for only a single process, some of which might stem from erroneous errors. Practically speaking, most processes at the point of the broadcast into global_daos_error are _most likely_ in a similar state, but to be safe in calling MPI_Finalize(), etc., it's best to make sure.
> One such out-of sync issue that could possibly arise, though unlikely, would be in our call to duns_resolve_path.
> Let's say we have 2 processes. One of them calls duns_resolve_path and succeeds. Before the other process makes the call, the UNS path is removed from the DAOS system, and then the 2nd process makes the call, and fails. Now the processes are in different states.
> Another such issue might be with mounting, since we could possibly run out of mountpoints for only one process.
> This is where my reasoning for "local_daos_error" comes from, among other erroneous network errors that might crop up.

Makes sense.  Good thinking.

Reply:
Oh yes... I'll need to rework that portion.

Reply:
This kind of check comes up a lot.  This utility function comes in handy:
https://github.com/hpc/mpifileutils/blob/a9cd09d6a4412b5a5e87419c8423d6d79f32caab/src/common/mfu_util.c#L1096

You can use it like `` if (! mfu_alltrue(fd >= 0, MPI_COMM_WORLD)) ``

Reply:
This is the full function:
https://github.com/daltonbohning/mpifileutils/blob/6f4a6d3252b720c770ed2bcb64c99dac3ae9f2bf/src/common/mfu_util.c#L1125-L1133

Then, I would use it like `if ! (mfu_alltrue(!local_daos_error, MPI_COMM_WORLD))`

But this function uses `MPI_INT`, `MPI_LAND`, whereas mine was using `MPI_C_BOOL`, `MPI_LOR`.
Do you know if there would be any performance difference here? I.e. perhaps MPI has an optimized implementation for `MPI_C_BOOL`, `MPI_LOR` vs `MPI_INT`, `MPI_LAND`?

Reply:
I actually found another hole in the logic here, where `same_pool` could potentially be mixed if a process has an error, and so the daos_connect call might be mixed. I ended up putting another reduction on `local_daos_error` before calling `daos_connect`. Even if refactored, I couldn't think of another way around this that was foolproof.

Reply:
> 
> 
> This is the full function:
> https://github.com/daltonbohning/mpifileutils/blob/6f4a6d3252b720c770ed2bcb64c99dac3ae9f2bf/src/common/mfu_util.c#L1125-L1133
> 
> Then, I would use it like `if ! (mfu_alltrue(!local_daos_error, MPI_COMM_WORLD))`
> 
> But this function uses `MPI_INT`, `MPI_LAND`, whereas mine was using `MPI_C_BOOL`, `MPI_LOR`.
> Do you know if there would be any performance difference here? I.e. perhaps MPI has an optimized implementation for `MPI_C_BOOL`, `MPI_LOR` vs `MPI_INT`, `MPI_LAND`?
> Or even just a difference between only the datatypes? Or only the `LAND` vs `LOR`? _**Theoretically**_, `MPI_C_BOOL`, `MPI_LOR` _**could**_ be implemented in a more efficient manner.

Yes, that's right.  One could implement an LAND/LOR allreduce on a bool using a single bit versus the 32-bits an int type would require.  It should be fine to modify the mfu_alltrue function to take a bool as its input if you're up for it.

Reply:
Sure, I'll do that (after a quick sanity check to everything calling mfu_alltrue).